src/rp66.cpp

```diff
@@ -0,0 +1,362 @@
+            this->read_header();
+
+            /* might be EOF, or even empty records, so re-start  */
+            continue;
+        }
+
+        assert(this->current.remaining >= 0);
+        std::int64_t n;
+        const auto to_read = std::min(len, this->current.remaining);
+        const auto err = this->fp->readinto(dst, to_read, &n);
+
+        this->current.remaining -= n;
+        bytes_read      += n;
+        dst = advance(dst, n);
+
+        if (err == LFP_OKINCOMPLETE)
+            return bytes_read;
+
+        if (err == LFP_EOF and this->current.remaining > 0) {
+            const auto msg = "rp66: unexpected EOF when reading Visible Record "
+                             "- got {} bytes from the current record,  "
+                             "expected there to be {} more";
+            throw unexpected_eof(fmt::format(msg, n, this->current.remaining));
+        }
+
+        if (err == LFP_EOF and this->current.remaining == 0)
+            return bytes_read;
+
+        assert(err == LFP_OK);
+
+        if (n == len)
+            return bytes_read;
+        /*
+         * The full read was performed, but there's still more requested - move
+         * onto the next segment. This differs from when read returns OKINCOMPLETE,
+         * in which case the underlying stream is temporarily exhausted or blocked,
+         * and fewer bytes than requested could be provided.
+         */
+
+        len -= n;
+    }
+}
+
+void rp66::parse_header() noexcept (false) {
+    assert(this->markers.empty()                    or
+           this->current     == this->markers.end() or
+           this->current + 1 == this->markers.end());
+
+    std::int64_t n;
+    unsigned char b[header::size];
+    auto err = this->fp->readinto(b, sizeof(b), &n);
+    switch (err) {
+        case LFP_OK: break;
+
+        case LFP_OKINCOMPLETE:
+            throw protocol_failed_recovery(
+                "rp66: incomplete read of Visible Record Header, "
+                "recovery not implemented"
+            );
+        case LFP_EOF:
+            /*
+             * The end of the *last* Visible Record aligns perfectly with
+             * EOF as there are no trailing bytes. Because EOF are typically
+             * not recorded before someone tries to read *past* the end, its
+             * perfectly fine to exhaust the last VR without EOF being set.
+             */
+            if (n == 0 && not this->markers.empty())
+                return;
+            else {
+                const auto msg = "rp66: unexpected EOF when reading header "
+                                 "- got {} bytes";
+                throw protocol_fatal(fmt::format(msg, n));
+            }
+
+
+        default:
+            throw not_implemented(
+                "rp66: unhandled error code in parse_header"
+            );
+    }
+
+    // Check the makefile-provided IS_LITTLE_ENDIAN, or the one set by gcc
+    #if (IS_LITTLE_ENDIAN || __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__)
```

Comment:
Something crossed my mind:
Do we even have IS_LITTLE_ENDIAN macro?
It seems like makefile sets only IS_BIG_ENDIAN.

Reply:
IS_LITTLE_ENDIAN is set by gcc, but not all compilers set it.

Reply:
Relic, I suppose, I think both used to be set. We should keep the check in code, and add it back to the makefiles for good measure.

Reply:
OK, whatever you say :smile: 
I though compilers set only ORDER_LITTLE_ENDIAN/ORDER_BIG_ENDIAN, but this line in makefile sets IS_BIG_ENDIAN
```
target_compile_definitions(lfp
    PRIVATE
        $<$<BOOL:${LFP_BIG_ENDIAN}>:IS_BIG_ENDIAN>
)
```
based on some `include(TestBigEndian)` which I don't know where is coming from

Reply:
__BYTE_ORDER__ is set by (some) compilers, the IS_LITTLE_ENDIAN is "ours".

include(TestBigEndian) is a cmake builtin.